src/ophyd_async/core/utils.py

```diff
@@ -77,6 +83,40 @@ def __str__(self) -> str:
         return self.format_error_string(indent="")
 
 
+@dataclass(frozen=True)
+class WatcherUpdate(Generic[T]):
+    """A dataclass such that, when expanded, it provides the kwargs for a watcher"""
+
+    name: str
+    current: T
+    initial: T
+    target: T
+    units: str
+    precision: float
+    fraction: float
+    time_elapsed: float
+    time_remaining: float
```

Comment:
Looking at https://github.com/bluesky/bluesky/blob/f03451a7e188df867df7fc86c70e3efe65a02709/src/bluesky/utils/__init__.py#L1301-L1343 then we should probably make these optional, then we don't have to calculate them every time
```suggestion
    fraction: Optional[float] = None
    time_elapsed: Optional[float] = None
    time_remaining: Optional[float] = None
```

Reply:
according to https://blueskyproject.io/bluesky/hardware.html#watch it kind of sounds like all/many of them should be optional

Reply:
This is true, but I suspect the minimum needed for a progress bar is `current`, `initial` and `target` as per https://github.com/bluesky/bluesky/blob/f03451a7e188df867df7fc86c70e3efe65a02709/src/bluesky/utils/__init__.py#L1315C40-L1315C64 so if people have opted into a `WatchableAsyncStatus` then they should at least provide those (otherwise they should just use `AsyncStatus`). The rest should probably be optional as you say.